src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs

```diff
@@ -48,229 +51,124 @@ private void Commit(PresentationItem item, Model model, char? commitChar)
-                            }
-
-                            var caretPoint = this.TextView.GetCaretPoint(this.SubjectBuffer);
-                            var virtualCaretPoint = this.TextView.GetVirtualCaretPoint(this.SubjectBuffer);
-
-                            if (caretPoint.HasValue && virtualCaretPoint.HasValue)
-                            {
-                                // TODO(dustinca): We need to call a different API here. TryMoveCaretToAndEnsureVisible might center within the view.
-                                this.TextView.TryMoveCaretToAndEnsureVisible(new VirtualSnapshotPoint(caretPoint.Value));
-                            }
-
-                            caretPoint = this.TextView.GetCaretPoint(this.SubjectBuffer);
-
-                            // Now that we're doing character level diffing, we need to move the caret to the end of 
-                            // the span being replaced. Otherwise, we can replace M|ai with Main and wind up with 
-                            // M|ain, since character based diffing makes that quite legit.
-                            if (caretPoint.HasValue)
-                            {
-                                var endInSubjectBuffer = this.TextView.BufferGraph.MapDownToBuffer(currentSpan.End, PointTrackingMode.Positive, caretPoint.Value.Snapshot.TextBuffer, PositionAffinity.Predecessor);
-                                if (caretPoint.Value < endInSubjectBuffer)
-                                {
-                                    this.TextView.TryMoveCaretToAndEnsureVisible(new SnapshotPoint(currentSpan.Snapshot.TextBuffer.CurrentSnapshot, currentSpan.End.Position));
-                                }
-                            }
-
-                            textEdit.Replace(currentSpan, newText);
-                        }
+                    // Now actually make the text change to the document.
+                    using (var textEdit = this.SubjectBuffer.CreateEdit(EditOptions.None, reiteratedVersionNumber: null, editTag: null))
+                    {
+                        var adjustedNewText = AdjustForVirtualSpace(textChange);
 
+                        textEdit.Replace(mappedSpan.Span, adjustedNewText);
                         textEdit.Apply();
                     }
 
                     // adjust the caret position if requested by completion service
-                    if (triggerSnapshotChange.NewPosition != null)
+                    if (completionChange.NewPosition != null)
                     {
-                        var target = new SnapshotPoint(this.SubjectBuffer.CurrentSnapshot, triggerSnapshotChange.NewPosition.Value);
-                        this.TextView.TryMoveCaretToAndEnsureVisible(target);
+                        TextView.Caret.MoveTo(new SnapshotPoint(
+                            this.SubjectBuffer.CurrentSnapshot, completionChange.NewPosition.Value));
```

Comment:
The contract of 'COmpletionChange' is that you give the NewPosition wrt to the final document after your text change is applied.